### PR TITLE
Replace regex with Secret type and remarshal config to hide secrets

### LINKED ISSUE #2722: Prometheus doesn't elide all secrets in config
The way we're hiding secrets in configs is based on a regex that hides the secret as long as it's on the same line as the field name.

This breaks if the secret is spread across multiple lines, such as is common with SSL certs.

We should either hide secrets in all cases, or in no cases to avoid giving a false sense of security.

The proper way to do this would be to re-marshall the YAML with secrets elided. This will however lose all comments, as there doesn't appear to be a Go YAML library that preserves comments.
On the minor plus side we'd be more likely to notice certain rare YAML formatting mistakes.

### PR DESCRIPTION
**Pull request by Conorbro**
Fixes #2722

### CODE REVIEW

File: cmd/prometheus/config_test.go [hunk 1]
```diff
@@ -98,7 +102,7 @@ func TestParseAlertmanagerURLToConfig(t *testing.T) {
 					i, test.username, acfg.HTTPClientConfig.BasicAuth.Username)
 			}
 
-			if test.password != acfg.HTTPClientConfig.BasicAuth.Password {
+			if config.Secret(test.password) != acfg.HTTPClientConfig.BasicAuth.Password {
```
Comment (brian-brazil) [inline cmd/prometheus/config_test.go]: it'd be better to change the type of the struct

File: config/config.go [hunk 2]
```diff
@@ -219,6 +219,27 @@ type Config struct {
 	original string
 }
 
+// Secret special type for storing secrets
```
Comment (brian-brazil) [inline config/config.go]: full stop at the end of comments

File: config/testdata/conf.good.yml [hunk 3]
```diff
@@ -67,7 +67,7 @@ scrape_configs:
 
   basic_auth:
     username: admin_name
-    password: admin_password
+    password: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```
Comment (brian-brazil) [inline config/testdata/conf.good.yml]: I'd say make this multi-line, and shorter.

Comment (brian-brazil) [inline config/testdata/conf.good.yml]: It'd also be good to test the other fields

File: config/config.go [hunk 4]
```diff
@@ -219,6 +219,27 @@ type Config struct {
...
+type Secret string
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface for Secrets.
+func (s *Secret) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type plain Secret
+	err := unmarshal((*plain)(s))
```
Comment (grobie) [inline config/config.go]: Why not just `return unmarshal((*plain)(s))`.

File: config/config.go [hunk 5]
```diff
@@ -219,6 +219,27 @@ type Config struct {
...
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// MarshalYAML implements the yaml.Marshaler interface for Secrets.
+func (s Secret) MarshalYAML() (interface{}, error) {
+	if s != "" {
+		return "XXXXX", nil
```
Comment (grobie) [inline config/config.go]: I wonder whether we should use a placeholder which makes it more clear that we changed the value, like `<hidden>` or so.

Comment (fabxc) [inline config/config.go]: Maybe "<redacted>" is better?

Comment (brian-brazil) [inline config/config.go]: Something like `<secret>`  would be clearer

File: config/config_test.go [hunk 6]
```diff
@@ -538,7 +539,7 @@ func TestLoadConfig(t *testing.T) {
 
 	// String method must not reveal authentication credentials.
 	s := c.String()
-	if strings.Contains(s, "admin_password") {
+	if strings.Contains(s, "admin_password") || strings.Contains(s, "xxxxxxxx") {
```
Comment (fabxc) [inline config/config_test.go]: You can directly check for the replacement value, no?

File: config/testdata/conf.good.yml (hunk 3 above)
Comment (grobie) [inline config/testdata/conf.good.yml]: What's the intention of this change?

File: config/testdata/conf.good.yml [hunk 7]
```diff
@@ -67,7 +67,7 @@ scrape_configs:
...
```
Comment (fabxc) [inline config/testdata/conf.good.yml]: What was wrong with "admin_password" though :)?

File: config/config.go [hunk 8]
```diff
@@ -284,14 +300,13 @@ func (c Config) String() string {
...
-			return fmt.Sprintf("<error creating config string: %s>", err)
-		}
-		s = string(b)
 	}
-	return patAuthLine.ReplaceAllString(s, "${1}<hidden>")
+	b, err := yaml.Marshal(c)
+	if err != nil {
+		return fmt.Sprintf("<error creating config string: %s>", err)
+	}
+	s = string(b)
```
Comment (brian-brazil) [inline config/config.go]: you can return this without the local variable.

File: config/config_test.go [hunk 9]
```diff
@@ -538,7 +540,9 @@ func TestLoadConfig(t *testing.T) {
 
 	// String method must not reveal authentication credentials.
 	s := c.String()
-	if strings.Contains(s, "admin_password") {
+	secretRe := regexp.MustCompile("<secret>")
+	matches := secretRe.FindAllStringIndex(s, -1)
+	if len(matches) < 5 {
 		t.Fatalf("config's String method reveals authentication credentials.")
```
Comment (brian-brazil) [inline config/config_test.go]: You might also want to test that the original passwords aren't present. That'll be a bit more resilient to someone adding new fields and not updating the number.

**Comment by Conorbro** [comment]
Updated this PR with better testing and other suggestions from the various reviews.

File: config/config_test.go [hunk 10]
```diff
@@ -538,9 +540,18 @@ func TestLoadConfig(t *testing.T) {
 
 	// String method must not reveal authentication credentials.
 	s := c.String()
-	if strings.Contains(s, "admin_password") {
+	secrets := []string{"admin\npassword", "avalidtoken", "mysecret", "nAdvAK2oBuVym4IXix", "mypassword"}
```
Comment (brian-brazil) [inline config/config_test.go]: Can we just stick with one value?

**Comment by Conorbro** [comment]
@brian-brazil

**Comment by arvenil** [comment]
The biggest issue with this approach is that now it's impossible to use those structs to save/load config as it gets saved without password. Would be nice to have possibility to disable this.

**Pull request merged**